### ISSUE TITLE
regression in recent jax release

### ISSUE THREAD

**Issue by proteneer** [google/jax#3407]
I was trying to build a set of benchmarks but unfortunately upon upgrading jax I discovered that our tests are now producing NaNs in the reference jax implementation. Upon bisecting:

Last working version:

jax==0.1.61
jaxlib==0.1.41

First failing version:

jax==0.1.62
jaxlib==0.1.42

I can try and write up a more concise repro but it will take me some time - just wondering if anyone else noticed this.

**Comment by proteneer** [comment]
See self-contained repro:

```
import jax.numpy as np
import jax

def f(x):
    ri = np.expand_dims(x, 0)
    rj = np.expand_dims(x, 1)
    dxdydz = np.power(ri - rj, 2)
    dij = np.sqrt(np.sum(dxdydz, axis=-1))

    keep_mask = 1 - np.eye(x.shape[0]) # keep everything except diagonals
    dij = np.where(keep_mask, dij, np.zeros_like(dij))
    eij = np.where(keep_mask, 1/dij, np.zeros_like(dij))
    return np.sum(eij)

x = np.array([
    [1.0, 2.0, 3.0],
    [0.2, 4.0, 6.0]
])

print(f(x))
df = jax.grad(f, argnums=(0,))
print(df(x)) # nans depending on version
```

**Comment by mattjj** [comment]
Thanks for raising this, @proteneer !

As crazy as it sounds, this regression was _intentional_. See [this comment on #1052](https://github.com/google/jax/issues/1052#issuecomment-514083352) and the PR message for #2447. It's quite subtle.

The upshot is that we need to write these kinds of `np.where` guards before _and_ after computations that might involve inf or nan Jacobians. Concretely, I _think_ I got this right:

```python
import jax.numpy as np
import jax

def f(x):
    ri = np.expand_dims(x, 0)
    rj = np.expand_dims(x, 1)

    dxdydz = np.power(ri - rj, 2)

    keep_mask = 1 - np.eye(x.shape[0])
    dij_sq = np.sum(dxdydz, axis=-1)
    dij_sq = np.where(keep_mask, dij_sq, 0)  # new!
    dij = np.sqrt(dij_sq)

    dij = np.where(keep_mask, dij, 0)
    eij = np.where(keep_mask, 1/dij, 0)
    return np.sum(eij)

x = np.array([
    [1.0, 2.0, 3.0],
    [0.2, 4.0, 6.0]
])

print(f(x))
df = jax.grad(f, argnums=(0,))
print(df(x)) # nans depending on version
```

WDYT?

**Comment by proteneer** [comment]
Thanks for getting back to me so quickly - and apologies for missing the discussion. Just so I understand, there are two sources of singularity:

1) 1/dij
2) sqrt(dij_sq)

Both of these needs to have the mask guard *before* and *after*? If so let me report back and see if its this works in general. We have a *lot* places where we need to do this

**Comment by proteneer** [comment]
Resolved - thanks for the pointers.
